File: Medium/CourseSchedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
 var canFinish = function(numCourses, prerequisites) {
    
    let adjList = {};
    for (let i = 0; i < numCourses; i++) adjList[i] = [];
    for (let pre of prerequisites) {
        const crs1 = pre[0];
        const crs2 = pre[1];
        adjList[crs1].push(crs2);
    }
    
    let visited = new Set();
    let dfs = function (course) {
        if (visited.has(course)) return false;
        if (adjList[course] === []) return true;
        
        visited.add(course);
        for (let neighbor of adjList[course]) {
            if (dfs(neighbor) === false) return false;
        }
        visited.delete(course);
        adjList[course] = [];
        return true;
    }
    
    for (let i = 0; i < numCourses; i++) {
        if (dfs(i) === false) return false;
    }
    
    return true;
};

/*

    DFS - check for cycle

    Time Complexity: O(v + e)
    Space Complexity: O(v + e)

*/

File: C#/Medium/CourseSchedule.cs
public class Solution {
    public bool CanFinish(int numCourses, int[][] prerequisites)
    {
        // build adjacency list
        var adjList = new Dictionary<int, List<int>>();
        for (int i = 0; i < numCourses; i++) adjList.Add(i, new List<int>());
        foreach (int[] prereq in prerequisites)
        {
            int course1 = prereq[0];
            int course2 = prereq[1];
            adjList[course1].Add(course2);
        }

        var visited = new HashSet<int>();
        bool dfs(int course)
        {
            if (visited.Contains(course)) return false;
            visited.Add(course);

            foreach (int neighbor in adjList[course])
            {
                if (!dfs(neighbor)) return false;
            }

            adjList[course].Clear(); // optimization
            visited.Remove(course);
            return true;
        }

        // check for a cycle starting from each course
        for (int i = 0; i < numCourses; i++)
        {
            if (!dfs(i)) return false;
        }

        return true;
    }
}

/*

    DFS - check for cycle

    Time Complexity: O(v + e)
    Space Complexity: O(v + e)

*/
